contents/security/cryptography.md

```diff
@@ -0,0 +1,81 @@
+
+* The key must be truely random, not pseudorandomly generated, and must never be reused.
+* The key must be securely distributed, and be at least as long as the message being generated. For example, to send a 10gb file to someone encrypted with the one-time pad requires sending 10gb of key material + 10gb of ciphertext = 20gb.
+
+In practice, we do not require perfect secrecy, since attackers have limited computational power. Hence, all other ciphers are only secure against such attackers.
+
+Modern stream ciphers approximate the operation of the one-time pad. 
+A short key (say 256 bits) is used to seed a [cryptographically secure pseudorandom number generator](https://en.wikipedia.org/wiki/Cryptographically_secure_pseudorandom_number_generator), which is used to generate the keystream for both encryption and decryption.
+They are more practical, as the key the communicating parties need to share is much shorter.
+
+Stream ciphers are used for their efficiency, ease of implementation in hardware, and when the length of the plaintext is unpredictable.
+However, block ciphers are more widely used than stream ciphers. In some modes of operation, they can be used like stream ciphers, reducing the need for dedicated stream ciphers.
+
+[RC4](https://en.wikipedia.org/wiki/RC4) is the most widely used stream cipher. Though its use is now discouraged. The [eSTREAM project](http://www.ecrypt.eu.org/stream/) is a research effort to develop state-of-the-art stream ciphers.
+
+### Block Ciphers
+
+Unlike stream ciphers, which operate on individual bits, block ciphers operate on an entire block of bits at a time. In practice, the size of each block is 64 or 128 bits.
+
+Shannon introduced 2 primitives, which modern block ciphers are built on.
+
+* Confusion: an operation wwhich obscures the relationship between key and ciphertext. This is usually done by substitution.
+* Diffusion: an operation which hides statistical properties in the plaintext by spreading the influence of a plaintext bit over many ciphertext bits. For example, the DES cipher achieves this by bit Permutations.
+
+Ciphers that use only one of these operations are insecure. For example, the insecure Caesar cipher only uses confusion. But strong ciphers can be built by using both confusion and diffusion - these are called product ciphers.
+
+* This [article](https://graquantum.com/blog/deciphering-encryption-des-block-cipher/) explains how the  DES cipher works, Feistel networks, s-boxes and p-boxes. Though DES is no longer secure, its design has inspired many ciphers. A still secure variant, triple DES, is popular in legacy applications.
+* The Advanced Encryption Standard (AES) is the most popular symmetric cipher today. It is used by the US government, many protocols such as TLS, WPA2-AES and SSH. This [article](https://graquantum.com/blog/deciphering-encryption-aes-block-cipher/) explains how AES works without going too much into the mathematical details.
```

Comment:
`It is used by the US government, many protocols such as TLS, WPA2-AES and SSH` I dont understand the comma